### ISSUE TITLE
Expose Central Scheduler metrics for easy ingestion

### ISSUE THREAD

**Issue by chrispalmer** [spotify/luigi#2036]
It would be extremely helpful to be able to get some metrics from the Central Scheduler via an RPC call. Will the visualizer gives you a view of what has been happening recently, it only shows a fixed amount of data (depending on your `remove-delay` config setting). Being able to continuously collect metrics would enable us to track overall health of our pipelines and the central scheduler.

Our particular use case would be to pull those metrics with Prometheus on a regular basis (although personally I'm still trying to understand how Prometheus works). But a more general solution not tailored to a particular metric collection tool, would still be helpful.

I've started exploring this and I'm still getting to grips with exactly what state the central scheduler stores. So I don't know what the complete list of potential metrics might be, but for starters they could include things like:

- Number of workers connected (per host)
- Number of tasks in each state (pending, running, failed etc) (per host where applicable)
- Average runtime of completed tasks

**Comment by chrispalmer** [comment]
I'd be very happy to work on adding this solution, as I've wanted to find something I could contribute back to this project for a while. Figured there might be some discussion about the correct way to build this.

**Comment by colemanja91** [comment]
This is a cool idea. I've been using Prometheus for about 7 months now; I was planning to add success/failure handlers to my tasks, but this would make it easier. I might also be interested in exploring metric collection for New Relic, as there's a chance I'll be migrating there in the near future. 

Output for Prometheus might look something like this: 
```
# HELP luigi_tasks_failed Count of tasks failed
# luigi_tasks_failed gauge
luigi_tasks_failed{task="task1"} 5
```

**Comment by Tarrasch** [comment]
I would love this too. I think you could start with a draftish PR (perhaps without tests) and ask for comments. Then you could iterate on the feedback from that and later turn it into a proper PR.

Do you see this as two parts? One to create the `/api/statistics` call and one to create a "Statistics" tab in the visualizer UI?

**Comment by chrispalmer** [comment]
One of the challenges with this is how much this wants to be integrated into the scheduler code. In my mind you could go one of two ways

1. Have an endpoint that when accessed, calculates/collects all the desired metrics and reports them back. In this case the endpoint would probably just get passed a reference to the scheduler, and it would go inspect `scheduler._state` and count various things.
2. Have the statistics being constantly collected and the endpoint simply reports back the collected metrics. This would require a persistent object of some sort to collect the metrics in and then add calls to the scheduler code to update various metrics as events happen.

A somewhat related choice, is if we would want to use Counters or Gauges for some of the metrics. Counters would mean that the metrics are collected over the lifetime of the central scheduler process, but you lose some information when you restart the server.

@colemanja91 Finding the right balance between metric names and labels is going to be interesting. I'd be tempted to have the task status be a label on the metric. I'd also fear that having the task family as a label would prove to be too high cardinality for effective use.

@Tarrasch I'm going to keep playing and maybe that will produce a draft PR of some sort. I definitely see there being an endpoint that exposes the metrics to what ever system wants to consume them. I'm not sure if it makes sense to add a tab to the visualizer, as I would expect the output to not be easily human readable. Or even if it is human readable the really interesting use is to see changes in the metrics over time. We'll see though.

**Comment by colemanja91** [comment]
@chrispalmer I'm not as familiar with the scheduler code, but it seems there's a dependency in there between the two approaches outlined and the question of Counters vs Gauges. 
If you calculate the metrics when the endpoint is accessed, then it seems intuitive to me that we would represent the metrics at that moment, therefore Gauges. 
Likewise, if the stats are constantly collected, then Counters. 

Totally agree on the metric names vs labels question. Was just copying that output as an example :) 
The monitoring practice I tend to follow is that any alert triggered by monitoring should be immediately actionable; hence, if I received some alert via Prometheus, I would want that alert to inform me which task had failed so that I could get to work debugging. 
However, using the Prometheus UI to just view monitor data would probably be way too overwhelming.

**Comment by chrispalmer** [comment]
@colemanja91 Yes, I think you are correct in terms of the two approaches. I think I have a preference for the constant collection and using Counters.

One potential approach that I think might be interesting is to actually hook into the event handlers logic that already exists. The scheduler would just trigger events when they happen, and you would have something `contrib.prometheus` or `contrib.newrelic` modules that register a number of event handlers and could deal with them as they want. Which of those modules actually gets imported would be controlled by a configuration setting.

I think that doesn't couple the scheduler code too tightly with any one metric collection tool, and provides lots of flexibility.  The only thing that wouldn't be possible is adding your own custom events to the scheduler, but I'm not sure that there is any reasonable solution to that problem (or if it is really a problem at all)

**Comment by chrispalmer** [comment]
Just submit a [pull request](https://github.com/spotify/luigi/pull/2044) for my work in progress. I've gone back and forth on various implementation details, and hope I've settled on a good one.

**Comment by chrispalmer** [comment]
@Tarrasch I don't know if you've had a chance to look at the PR yet, but I've been thinking about it some more. I had a fundamental question and I'm not sure who the best person to answer it is.

My initial implementation puts most of the work in the `SimpleTaskState` class because I found that was the easiest place to detect task status changes. However, I wonder if that is the right place for it, or if it should be in the `Scheduler` class itself.

That would require a little bit more code I think but feels like a better place for this functionality. Do you have any thoughts on that?

**Comment by aba182** [comment]
😢
